### PR TITLE
#554: Lead/lag window function with offset and default value arguments

### LINKED ISSUE #554: implement lead and lag with 2nd and 3rd argument
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

follow up with #553 , we can additionally support 2nd and 3rd argument

**Describe the solution you'd like**

Similar to postgresql function definitions

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jgoday**
# Which issue does this PR close?
#Closes #554.

 # Rationale for this change
Implement offset and default value optional arguments in lead/lag window functions.

# What changes are included in this PR?
Changes in lead/lag pub fn signatures.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jgoday** [comment]
I have a few doubts:

1) **Lag/Lead** signatures has to be changed to support 2nd/3nd optional arguments, is this correct or am I misunderstanding the use of **Signature::OneOf** ?
```rust
Signature::OneOf(vec![
    Signature::Any(1),
    Signature::Any(2),
    Signature::Any(3),
])
```

2) If previous point is correct, then evaluation of **Signature::OneOf** in **type_coercion::get_valid_types** has to be changed to something like this 
```rust
Signature::OneOf(types) => {
    let mut r = vec![];
    for s in types {
        // we must ensure that one of the signatures is valid against the current_types
        if let Ok(valid_types) = get_valid_types(s, current_types) {
            r.extend(valid_types);
        }
    }
    r
}
```
original code is currently forcing all options within **Signature::OneOf** to be valid
```rust
Signature::OneOf(types) => {
    let mut r = vec![];
    for s in types {
        r.extend(get_valid_types(s, current_types)?);
    }
    r
}
```

3) I'd like to test **create_built_in_window_expr** with the optional arguments on and off, to match against the **lead/lag** signatures,
but How can I test the resulting **WindowShift** struct?
Right now, I can only try that the presence or absence of this optional arguments don't panic in **test_create_window_exp_lead_no_args/test_create_window_exp_lead_with_args**

4) Had to copy and change the original **arrow::compute::kernels::window::shift** to allow filling the start/end nulls with the default value (this new fn is called **shift_with_default_value** inside lead_lag crate)

**Comment by alamb** [comment]
FYI @Jimexist  -- do you have some time to review this code?

File: datafusion/src/physical_plan/expressions/lead_lag.rs [hunk 1]
```diff
@@ -98,20 +106,58 @@ impl BuiltInWindowFunctionExpr for WindowShift {
...
+    }
 }
 
 impl PartitionEvaluator for WindowShiftEvaluator {
     fn evaluate_partition(&self, partition: Range<usize>) -> Result<ArrayRef> {
         let value = &self.values[0];
         let value = value.slice(partition.start, partition.end - partition.start);
-        shift(value.as_ref(), self.shift_offset).map_err(DataFusionError::ArrowError)
+        if let Some(default_value) = self.default_value.clone() {
+            shift_with_default_value(&value, self.shift_offset, default_value)
```
Comment (Jimexist) [inline datafusion/src/physical_plan/expressions/lead_lag.rs:157]: in both branches you can just call this `shift_with_default_value`. if the default value isn't present, you can unwrap or provide a null scalar value

File: datafusion/src/physical_plan/type_coercion.rs [hunk 2]
```diff
@@ -131,7 +131,10 @@ fn get_valid_types(
         Signature::OneOf(types) => {
             let mut r = vec![];
             for s in types {
-                r.extend(get_valid_types(s, current_types)?);
+                if let Ok(valid_types) = get_valid_types(s, current_types) {
```
Comment (Jimexist) [inline datafusion/src/physical_plan/type_coercion.rs:134]: why is this necessary?

**Review by Jimexist — COMMENTED** [review]
this change looks generally okay, thanks for the effort!

I would suggest: replacing branching with one call and also add a case in the integration test folder, esp.:

1. out of bounds shift offset, 
2. negative bounds
3. default value being null

**Comment by Jimexist** [comment]
> Lag/Lead signatures has to be changed to support 2nd/3nd optional arguments, is this correct or am I misunderstanding the use of Signature::OneOf ?

that part was confusing to me as well. i believe a most elegant way is to create support for heterogenous support for function arguments 

> I'd like to test create_built_in_window_expr with the optional arguments on and off, to match against the lead/lag signatures,

i'm okay with integration tests that cover these

> Had to copy and change the original arrow::compute::kernels::window::shift to allow filling the start/end nulls with the default value (this new fn is called shift_with_default_value inside lead_lag crate)

changing arrow-rs repo is one way to achieve it although it'll needs two more weeks to be released. i'm okay with keeping the copied and modified version here.

File: datafusion/src/physical_plan/type_coercion.rs (hunk 2 above)
Comment (jgoday) [reply datafusion/src/physical_plan/type_coercion.rs:134]: Maybe I have misunderstood it, but in the original code, if one of the signatures within OneOf fails then the whole get_valid_types returns an error, isn't it ?

Comment (Jimexist) [reply datafusion/src/physical_plan/type_coercion.rs:134]: you can also use [`filter_map`](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#ignore-the-failed-items-with-filter_map) and then take first

Comment (jgoday) [reply datafusion/src/physical_plan/type_coercion.rs]: Thanks ! changed to filter_map.

File: datafusion/src/physical_plan/expressions/lead_lag.rs (hunk 1 above)
Comment (jgoday) [reply datafusion/src/physical_plan/expressions/lead_lag.rs]: Remove the conditional statement and changed to one branch with one call to shift_with_default_value and the scalar_value as an option

**Comment by jgoday** [comment]
> this change looks generally okay, thanks for the effort!
> 
> I would suggest: replacing branching with one call and also add a case in the integration test folder, esp.:
> 
>     1. out of bounds shift offset,
> 
>     2. negative bounds
> 
>     3. default value being null

Thank you @Jimexist.
I have followed your advices and changed the type_coercion::get_valid_types to use filter_map and switch the two conditional branches to one single call to shift_with_default_value. I will try to look the integration tests later today.

**Review by Jimexist — APPROVED** [review]
implementation looks great, 👍 thanks!

would be great to add integration test coverage as well before merge

File: datafusion/src/physical_plan/expressions/lead_lag.rs [hunk 3]
```diff
@@ -98,20 +105,58 @@ impl BuiltInWindowFunctionExpr for WindowShift {
...
     }
 }
 
 pub(crate) struct WindowShiftEvaluator {
     shift_offset: i64,
     values: Vec<ArrayRef>,
+    default_value: Option<ScalarValue>,
+}
+
+fn shift_with_default_value(
```
Comment (Jimexist) [inline datafusion/src/physical_plan/expressions/lead_lag.rs:119]: you can add a comment of todo to push this upstream to arrow-rs

File: datafusion/src/physical_plan/windows.rs [hunk 4]
```diff
@@ -110,13 +110,49 @@ fn create_built_in_window_expr(
...
         }
         BuiltInWindowFunction::Lead => {
             let coerced_args = coerce(args, input_schema, &signature_for_built_in(fun))?;
             let arg = coerced_args[0].clone();
             let data_type = args[0].data_type(input_schema)?;
-            Arc::new(lead(name, data_type, arg))
+            let shift_offset = coerced_args
+                .get(1)
+                .map(|v| {
+                    v.as_any()
```
Comment (Jimexist) [inline datafusion/src/physical_plan/windows.rs:140]: feel like you can have a helper function to extract argument as `Result<Literal>`

Comment (jgoday) [reply datafusion/src/physical_plan/windows.rs]: Ok, created get_scalar_value_from_args helper function.

File: datafusion/src/physical_plan/expressions/lead_lag.rs (hunk 3 above)
Comment (jgoday) [reply datafusion/src/physical_plan/expressions/lead_lag.rs:139]: Added a TODO comment

**Comment by jgoday** [comment]
@Jimexist I was trying to execute the integration tests manually,
executing test_psql_parity.py with the following sql
```
SELECT
  c8,
  LEAD(c8) OVER () next_c8,
  LEAD(c8, 10, 10) OVER() next_10_c8,
  LEAD(c8, 100, 10) OVER() next_overflow_c8,
  LAG(c8) OVER() prev_c8,
  LAG(c8, -2, 0) OVER() AS prev_2_c8,
  LAG(c8, -200, 10) OVER() AS prev_overflow_c8

FROM test
ORDER BY c8;
```

There was some errors: type coercing problems between default_value and the slice data_type
and between lead postgres (positive) and datafusion/shifting (negative) arguments.
After fixed it in the latest commit, previous example works fine.

But if I try to test agains some custom partition
```sql
SELECT
  LAG(c8) OVER(PARTITION BY c9 ORDER BY c9) AS prev_c8

FROM test
ORDER BY c8;
```
datafusion throws an error
```
Empty DataFrame
Columns: [Plan("Projections require unique expression names but the expression \"LAG(#test.c8) AS prev_c8\" at position 4 and \"LAG(#test.c8) PARTITION BY [#test.c9] ORDER BY [#test.c9 ASC NULLS FIRST] AS prev_c8\" at position 7 have the same name. Consider aliasing (\"AS\") one of them.")]

```
I will try to look at this later today if I can ...

**Comment by Jimexist** [comment]
if the first example works then it's fine. the second example might fail because of different reasons not related to this change.

**Comment by jgoday** [comment]
> if the first example works then it's fine. the second example might fail because of different reasons not related to this change.

Would it be better to have a different sql for each case (lead/lag out of bounds/negative/default being null) or is it ok to test all combinations in a single sql?

**Comment by alamb** [comment]
@Jimexist  do you think this PR is ready? Do you need help reviewing ?

**Comment by Jimexist** [comment]
> @Jimexist do you think this PR is ready? Do you need help reviewing ?

looks okay after rebasing.

**Comment by jgoday** [comment]
> > @Jimexist do you think this PR is ready? Do you need help reviewing ?
> 
> looks okay after rebasing.

Hi @Jimexist, just made the rebase from master

File: datafusion/src/physical_plan/expressions/lead_lag.rs [hunk 5]
```diff
@@ -176,6 +240,28 @@ mod tests {
...
+        test_i32_result(
+            lag(
+                "lead".to_owned(),
+                DataType::Int32,
+                Arc::new(Column::new("c3", 0)),
+                None,
+                Some(ScalarValue::Int32(Some(100))),
+            ),
+            vec![
+                Some(100),
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/lead_lag.rs:253]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me . Thank you @jgoday  and @Jimexist
